libs/counterfactuals/src/util/largeCounterfactualsView/getCounterfactualsScatter.ts

```diff
@@ -0,0 +1,224 @@
+// Licensed under the MIT License.
+
+import {
+  FluentUIStyles,
+  IGenericChartProps,
+  JointDataset
+} from "@responsible-ai/core-ui";
+import { WhatIfConstants } from "@responsible-ai/interpret";
+import { localization } from "@responsible-ai/localization";
+import { PointMarkerOptionsObject } from "highcharts";
+import { Dictionary } from "lodash";
+
+import { buildScatterTemplate } from "./buildScatterTemplate";
+
+export interface IDatasetExplorerSeries {
+  name?: string;
+  color: any;
+  data: IDatasetExplorerData[];
+  marker?: PointMarkerOptionsObject;
+}
+export interface IDatasetExplorerData {
+  x: number;
+  y: number;
+  customData: any;
+  template: string | undefined;
+}
+
+export function getCounterfactualsScatter(
+  xSeries: number[],
+  ySeries: number[],
+  indexSeries: number[],
+  jointDataset: JointDataset,
+  selectedPointsIndexes: number[],
+  chartProps?: IGenericChartProps,
+  customPoints?: Array<{ [key: string]: any }>
+): any[] {
+  const dataSeries: any = [];
+  const result = [];
+  // const customData = plotlyProps.data[0].customdata;
+  const xData = xSeries;
+  const yData = ySeries;
+
+  if (yData) {
+    yData.forEach((data, index) => {
+      dataSeries.push({
+        customData:
+          chartProps &&
+          buildScatterTemplate(
+            jointDataset,
+            chartProps,
+            xData?.[index],
+            data,
+            index,
+            indexSeries[index]
+          ),
+        marker: getMarker(selectedPointsIndexes, index),
+        x: xData?.[index],
+        y: data
+      });
+    });
+  }
+
+  dataSeries.map(
+    (d: {
+      customData: { Index: number };
+      marker: { fillColor: string; radius: number; symbol: string };
+    }) => {
+      const selectionIndex = selectedPointsIndexes.indexOf(d.customData.Index);
+      const color =
+        selectionIndex === -1
+          ? FluentUIStyles.fabricColorInactiveSeries
+          : FluentUIStyles.fluentUIColorPalette[selectionIndex];
+      return (d.marker = {
+        fillColor: color,
+        radius: 4,
+        symbol: selectionIndex === -1 ? "circle" : "square"
+      });
+    }
+  );
+
+  if (customPoints) {
+    const customPointsCustomData = getCustomPointCustomData(
+      customPoints,
+      jointDataset,
+      chartProps
+    );
+    customPoints.forEach((_cp, index) => {
+      dataSeries.push({
+        customData: customPointsCustomData[index],
+        marker: getCustomPointMarker(customPoints, index),
+        x: customPointsCustomData[index].rawXValue,
+        y: customPointsCustomData[index].rawYValue
+      });
+    });
+  }
+
+  result.push({
+    data: dataSeries
+  });
+  return result;
+}
+
+function getMarker(selectedPointsIndexes: number[], index: number): any {
+  const selectionIndex = selectedPointsIndexes.indexOf(index);
+  const color =
+    selectionIndex === -1
+      ? FluentUIStyles.fabricColorInactiveSeries
+      : FluentUIStyles.fluentUIColorPalette[selectionIndex];
```

Comment:
I'm seeing this logic in multiple places. Can we put that in a function?

Reply:
good catch, updated